src/app/MessageDef.h

```diff
@@ -0,0 +1,894 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+/**
+ *    @file
+ *      This file defines CHIP interaction model message parsers and encoders
+ *      Interaction model profile.
+ *
+ */
+
+#ifndef _CHIP_INTERACTION_MODEL_MESSAGE_DEF_H
+#define _CHIP_INTERACTION_MODEL_MESSAGE_DEF_H
+
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <util/basic-types.h>
+
+namespace chip {
+namespace app {
+/**
+ *  @brief
+ *    The CHIP interaction model message types.
+ *
+ *  These values are called out in CHIP Interaction Model: Encoding Specification
+ *
+ */
+enum
+{
+    kMsgType_SubscribeRequest      = 0x01,
+    kMsgType_ReadRequest           = 0x02,
+    kMsgType_ReportData            = 0x03,
+    kMsgType_WriteRequest          = 0x04,
+    kMsgType_WriteResponse         = 0x05,
+    kMsgType_InvokeCommandRequest  = 0x06,
+    kMsgType_InvokeCommandResponse = 0x07,
+};
+
+class ParserBase
+{
+public:
+    /**
+     *  @brief Initialize a TLVReader to point to the beginning of any tagged element in this request
+     *
+     *  @param [in]  aTagToFind Tag to find in the request
+     *  @param [out] apReader   A pointer to TLVReader, which will be initialized at the specified TLV element
+     *                          on success
+     *
+     *  @retval #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR GetReaderOnTag(const uint64_t aTagToFind, chip::TLV::TLVReader * const apReader) const;
+
+protected:
+    chip::TLV::TLVReader mReader;
+
+    ParserBase();
+
+    template <typename T>
+    CHIP_ERROR GetUnsignedInteger(const uint8_t aContextTag, T * const apLValue) const;
+
+    template <typename T>
+    CHIP_ERROR GetSimpleValue(const uint8_t aContextTag, const chip::TLV::TLVType aTLVType, T * const apLValue) const;
+};
+
+class ListParserBase : public ParserBase
+{
+protected:
+    ListParserBase();
+
+public:
+    // aReader has to be on the element of the array element
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
+
+    // aReader has to be at the beginning of some container
+    CHIP_ERROR InitIfPresent(const chip::TLV::TLVReader & aReader, const uint8_t aContextTagToFind);
+
+    CHIP_ERROR Next();
+    void GetReader(chip::TLV::TLVReader * const apReader);
+};
+
+class BuilderBase
+{
+public:
+    void ResetError();
+    void ResetError(CHIP_ERROR aErr);
+    CHIP_ERROR GetError() const { return mError; };
+    chip::TLV::TLVWriter * GetWriter() { return mpWriter; };
+
+protected:
+    CHIP_ERROR mError;
+    chip::TLV::TLVWriter * mpWriter;
+    chip::TLV::TLVType mOuterContainerType;
+
+    BuilderBase();
+    void EndOfContainer();
+
+    CHIP_ERROR InitAnonymousStructure(chip::TLV::TLVWriter * const apWriter);
+};
+
+class ListBuilderBase : public BuilderBase
```

Comment:
could we strike the `Base` suffix from all these, seeing as how every derived class is in its own namespace?

Reply:
not sure i get the points.
for example, InvokeCommand has Parser and Builder inside, and they inherit from ParserBase and BuilderBase, it seems I need Base for differentiation purpose.

namespace InvokeCommand {
class Parser : public ParserBase
class Builder : public BuilderBase
}

Reply:
I mean:

```
class ListBuilder : public Builder;
```

later:
```
namespace AttributePath {
class Builder : public chip::app::Builder
...
```

Reply:
Done, thanks